### PR TITLE
Added support for custom bulk selling options

### PR DESCRIPTION
**Pull request by BuckAMayzing**
DRAFT - not ready for merge yet

This logic is all coming from [More Bulk Sell](https://www.nexusmods.com/pathfinderwrathoftherighteous/mods/163?tab=files). The source is available there if you'd like to look. I've cleaned up a _lot_ of the existing UI code. I still need to test this in-game and am leaving it as a draft until then.

Done: settings persistence, UI, and game logic.

Todo: test game logic

Screenshots (state as of 5/9/2023):

image
image
image
image

### CODE REVIEW

File: ModKit/Utility/Dictionary/SerializableDictionary.cs [hunk 1]
```diff
@@ -6,18 +6,22 @@
 
 namespace ModKit.Utility {
     [XmlRoot("SerializableDictionary")]
-    public class SerializableDictionary<TKey, TValue> : Dictionary<TKey, TValue>, IXmlSerializable, IUpdatableSettings {
+    public class SerializableDictionary<TKey, TValue> : Dictionary<TKey, TValue>, IXmlSerializable, IUpdatableSettings
```
Comment (BuckAMayzing) [inline ModKit/Utility/Dictionary/SerializableDictionary.cs:9]: I did a thing in this class that I ended up not needing. Apparently it auto-formatted it. If you want me to revert, that shouldn't be a big deal.

File: ToyBox/ToyBox.csproj [hunk 2]
```diff
@@ -315,7 +318,7 @@
   </PropertyGroup>
   <ProjectExtensions>
     <VisualStudio>
-      <UserProperties info_1json__JsonSchema="https://json.schemastore.org/global.json" config_4bindings_1json__JsonSchema="https://json.schemastore.org/global.json" />
+      <UserProperties config_4bindings_1json__JsonSchema="https://json.schemastore.org/global.json" info_1json__JsonSchema="https://json.schemastore.org/global.json" />
```
Comment (BuckAMayzing) [inline ToyBox/ToyBox.csproj:321]: No clue what happened here or why. Looks like the order changed for some reason. I didn't make this change.

File: ToyBox/classes/MonkeyPatchin/Vendor/TrashLoot.cs [hunk 3]
```diff
@@ -0,0 +1,306 @@
+﻿using HarmonyLib;
```
Comment (BuckAMayzing) [inline ToyBox/classes/MonkeyPatchin/Vendor/TrashLoot.cs:1]: I wouldn't bother reviewing this file yet; it's untouched from the source mod. I'll be doing a lot of work to it.

The file location though - does that work for you?

File: ModKit/Utility/Dictionary/SerializableDictionary.cs (hunk 1 above)
Comment (cabarius) [reply ModKit/Utility/Dictionary/SerializableDictionary.cs:9]: Please don 't change formatting if you can avoid it.

File: ToyBox/ToyBox.csproj (hunk 2 above)
Comment (cabarius) [reply ToyBox/ToyBox.csproj:321]: I dunno why this happens.  ADDB hits the same issue.

File: ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+namespace ToyBox
+{
+    internal static class BulkSell
+    {
+        private static readonly BulkSellSettings _settings = Main.Settings.bulkSellSettings;
+
+        public static void OnGUI()
+        {
+            void bonusItemOptions(string itemTypeName, string bonusType, ref int enchantLevel, ref int stackSize)
```
Comment (cabarius) [inline ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs:16]: Please capitalize method names

File: ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            void damageTypeOptions<T>(SerializableDictionary<T, bool> settings) where T : Enum
+            {
+                settings.Keys
+                    .Select((type, index) => new { type, index })
+                    .GroupBy(g => g.index / 4)
+                    .ToList()
+                    .ForEach(group =>
+                    {
```
Comment (cabarius) [inline ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs:42]: please do brace on the same line as the construct.

File: ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs (hunk 4 above)
Comment (BuckAMayzing) [reply ToyBox/classes/MainUI/EnhancedUI/BulkSell.cs:16]: Even locally scoped functions like this?

File: ModKit/Utility/Dictionary/SerializableDictionary.cs (hunk 1 above)
Comment (BuckAMayzing) [reply ModKit/Utility/Dictionary/SerializableDictionary.cs:9]: Alright. Will revert.

**Comment by cabarius** [comment]
Code changes look fine.

Please configure your style settings to match those of ToyBox.  I thought the editor config in the project was set up correctly.   Two main issues 
1) Brace formatting.  Should look like this
```cs
        public static void OnGUI(EtudeStatus etudeStatus, object source) {
            Label("Etude Status: ".cyan(), 150.width());
            var etudeBP = etudeStatus.Etude;
            Label(etudeBP.name.orange(), 500.width());
            var etudeState = Game.Instance.Player.EtudesSystem.GetSavedState(etudeBP);
            var debugInfo = Game.Instance.Player.EtudesSystem.GetDebugInfo(etudeBP);
            22.space();
            using (VerticalScope()) {
                HelpLabel(debugInfo);
                Label($"status: ".cyan() + etudeState.ToString());
                Label("condition: ".cyan() + etudeStatus.CaptionString());
                Label("source: ".cyan() + source.ToString().yellow());
            } ```
2) Please follow the naming conventions in the mod.  I know the property naming is kind of messed up but trying to move to PascalCase for public properties and _member for private.  And please use PascalCase for Method names

File: .editorconfig [hunk 6]
```diff
@@ -138,3 +138,4 @@ dotnet_style_object_initializer = true:suggestion
 dotnet_style_collection_initializer = true:suggestion
 dotnet_style_prefer_simplified_boolean_expressions = true:suggestion
 csharp_new_line_before_catch = true
+csharp_new_line_before_open_brace = none
```
Comment (BuckAMayzing) [inline .editorconfig:141]: I added this to reflect the desired state. If you'd prefer that this become a separate PR so it gets its own commit, let me know.

File: ToyBox/classes/MainUI/EnhancedUI/PhatLoot.cs [hunk 7]
```diff
@@ -276,7 +290,7 @@ public class PhatLoot {
                            }
                        }
                    },
-                   () => {});
+                   () => { });
```
Comment (BuckAMayzing) [inline ToyBox/classes/MainUI/EnhancedUI/PhatLoot.cs:293]: Auto formatter made this and the trailing whitespace change. Since I had to touch the file anyway, I think it's worth leaving alone so it doesn't happen in the future.

File: ToyBox/ToyBox.csproj (hunk 2 above)
Comment (BuckAMayzing) [reply ToyBox/ToyBox.csproj]: ...and it changed it back. Weird.

**Review by cabarius — APPROVED** [review]
